front : refactor Locators to Use data-testid

Maymanaf opened the pull request:
Closes #12481

Maymanaf committed:
front: update project page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update study page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update scenario page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update route page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update simulation settings page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update rolling stock editor page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update rolling stock selector page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update operational studies page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update scenario timetable page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update simulation results pages model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update stdcm page model to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update stdcm sections to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: update paced train section to use data-testid

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Synar commented on front/tests/pages/operational-studies/simulation-settings-tab.ts:
```diff
@@ -9,14 +9,14 @@ class SimulationSettingsTab {
   private readonly marecoMarginSwitch: Locator;
 
-  private readonly codeCompoSelector: Locator;
+  private readonly speedLimitTagSelector: Locator;
```
Nice!

Maymanaf replied:
I've updated the description of the last commit

Synar reviewed (approved):
LGTM for the first 13 commits, great work! I really like moving to explicit test ids.

I don't quite understand the very last commit though. Could you explain the individual changes you performed in it?

Maymanaf committed:
front: fix flaky PDF parsing and redundant awaits in tests

- Replaced `new Uint8Array(buffer.buffer)` with `new Uint8Array(buffer)` to fix a DataCloneError when parsing PDF with pdfjs-dist.

- Removed redundant `await` in expect statements.

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Yohh commented on front/src/common/BootstrapSNCF/OptionsSNCF.tsx:
```diff
@@ -47,7 +48,11 @@ export default function OptionsSNCF({
-            <label className="options-btn font-weight-medium" htmlFor={`${name}${option.value}`}>
+            <label
+              className="options-btn font-weight-medium"
+              htmlFor={`${name}${option.value}`}
```
nit:
```suggestion
              htmlFor={name + option.value}
```

Synar replied:
I believe in the front we prefer format strings over addition, even for 'trivial' cases like this. And we have a eslint warning to replace additions by format strings.

So I would personally leave the format string as is.

Yohh replied:
mm ok fine

Yohh reviewed (approved):
just one nit, it's not related to your work, just a little clean so feel free to skip it if you want

Synar commented:
Thanks, much clearer now!

Nice catch on the buffet.buffer vs buffer issue, this one seemed quite technical.

Just one tiny issue, your last commit mentions :

> - Added assertions like `expect(...).toBeVisible()` to improve test robustness.

But I don't believe it does that

Maymanaf commented:
> Thanks, much clearer now!
> 
> Nice catch on the buffet.buffer vs buffer issue, this one seemed quite technical.
> 
> Just one tiny issue, your last commit mentions :
> 
> > * Added assertions like `expect(...).toBeVisible()` to improve test robustness.
> 
> But I don't believe it does that

The assertions were added in a previous [PR ](https://github.com/OpenRailAssociation/osrd/pull/12572#event-18812504352)  so the changes were lost after a rebase.
I have updated the commit name and description.

clarani commented on front/src/common/BootstrapSNCF/OptionsSNCF.tsx:
```diff
@@ -47,7 +48,11 @@ export default function OptionsSNCF({
+            <label
+              className="options-btn font-weight-medium"
+              htmlFor={`${name}${option.value}`}
+              data-testid={`${name}${option.value}-label`}
```
What about adding a `-` to improve lisibility ?
```suggestion
              data-testid={`${name}-${option.value}-label`}
```

clarani commented on front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorFormHelpers.tsx:
```diff
@@ -197,6 +197,7 @@ const RollingStockEditorParameterFormColumn = ({
               <SelectSNCF
                 sm
                 id={property.title}
+                data-testid={property.title}
```
Hum I think you don't need this line anymore since you're using the id ? (commit `front: update rolling stock editor page model to use data-testid`)
```suggestion
```

clarani commented on front/src/common/BootstrapSNCF/CheckboxRadioSNCF.tsx:
```diff
@@ -36,6 +37,7 @@ const RadioInput = (props: CheckboxRadioProps) => {
       <input
         {...inputProps}
         id={id}
+        data-testid={id}
```
You don't want to add `-checkbox` here ? (commit `front: update operational studies page model to use data-testid`)

clarani reviewed (commented):
Very nice refacto :tada:
I'm just thinking a bit, do you think it's good practice to add hard-coded datatest-ids in osrd-ui ?
Also, it would be nice to group all the changes related to ui in a single commit (and to have only these in it). It would be clearer for our external users

Maymanaf commented:
> Very nice refacto 🎉 I'm just thinking a bit, do you think it's good practice to add hard-coded datatest-ids in osrd-ui ? Also, it would be nice to group all the changes related to ui in a single commit (and to have only these in it). It would be clearer for our external users

I don't see any problem with adding the datatest-ids, they're useful for testing stability not just for our project but for any other app
And you're right about grouping all osrd-ui changes into a single commit, it would make the history much cleaner

Maymanaf committed:
front: add testIdPrefix prop across osrd-ui components for testability

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

Maymanaf committed:
front: refactor tests to use testIdPrefix-based components

Signed-off-by: maymanaf <med.aymen.naf@gmail.com>

clarani reviewed (approved):
LGTM :white_check_mark:
